rust/macros/module.rs

```diff
@@ -512,17 +515,31 @@ pub(crate) fn module(ts: TokenStream) -> TokenStream {
             static THIS_MODULE: kernel::ThisModule = unsafe {{ kernel::ThisModule::from_ptr(core::ptr::null_mut()) }};
 
             // Loadable modules need to export the `{{init,cleanup}}_module` identifiers
+            /// Tries to initialize the module and then unsynchronously sets `__MOD` to the initialized module.
+            /// Internally calls `__init()`.
+            ///
+            ///
+            /// # Safety
+            ///
+            /// This method must be called before any function of the kernel module can be used, and
+            /// must not be called unsynchronously with `cleanup_module()`, which unsets `__MOD`.
             #[cfg(MODULE)]
             #[doc(hidden)]
             #[no_mangle]
-            pub extern \"C\" fn init_module() -> kernel::c_types::c_int {{
+            pub unsafe extern \"C\" fn init_module() -> kernel::c_types::c_int {{
                 __init()
             }}
```

Comment:
Please note that we use `deny(unsafe_block_in_unsafe_fn)` like Gary mentioned, thus we are using a "dialect" of the language where an unsafe function does not imply the body is within an unsafe block. Thus the `__init()` call should have its own `unsafe` block (and `// SAFETY:` comment on top). Same for `__exit()` calls.

(Proc macros "eat" some of the lints at the moment, so this does not show up, but we should still have the correct code in case this changes in the future).